Using remat with host_callback is broken

This throws an error:

```
import jax
from jax.experimental import host_callback

def f1(x):
  x = host_callback.id_tap(print, x)
  return x

def f2(x, _):
  f1_r = jax.remat(f1)
  x = f1_r(x)
  return x, 1

def forward(x):
  x = jax.lax.scan(f2, x, None, length=2)
  return x[0]

jax.jit(jax.grad(forward))(1.2)
```

```
...
# /usr/local/lib/python3.7/dist-packages/jax/experimental/host_callback.py in _outside_call_partial_eval_rule(trace, *args, **params)
    985   primals, tangents = util.split_list(args, [nr_primals])
    986   c_primals_tapped, _ = util.split_list(consts, [nr_primals])
--> 987   assert all([c is not None for c in c_primals_tapped])
    988 
    989   prims, _ = params["arg_treedef"].unflatten(args)

AssertionError: 
```
It runs fine if the line `x = host_callback.id_tap(print, x)` is removed, or if no `lax.scan` is used.

--- a/jax/experimental/host_callback.py
+++ b/jax/experimental/host_callback.py
@@ -1215,28 +1215,31 @@ ad.primitive_jvps[outside_call_p] = _outside_call_jvp_rule
 
 
 def _outside_call_partial_eval_rule(trace, *args, **params):
-  # The args have been prepared by the id_tap_jvp_rule: primals, tangents
+  # partial eval is used after jvp and before transpose.
   transforms = params.get("transforms", ())
   if not transforms or transforms[-1] != ("jvp",):
     # We are not in the process of computing VJP
     return trace.default_process_primitive(outside_call_p, args, params)
 
+  # The args have been prepared by the id_tap_jvp_rule: primals, tangents. The
+  # result is a pair of the primal outputs and output tangents.
+  # One invariant that JAX requires is that if the primals arguments are known
+  # then the primal outputs must be known. So, if the primal arguments are known
+  # and some of the tangents are unknown, then we must split the tap into
+  # one for the primals (thus the output will be considered known), and a
+  # separate tap for the tangents.
   assert "has_token" not in params
   if not params["identity"]:
     raise NotImplementedError("differentiation rules are implemented only for id_tap, not for call.")
 
   assert len(args) % 2 == 0
   nr_primals = len(args) // 2
-
-  consts = [t.pval.get_known() for t in args]
-  if all(c is not None for c in consts):
-    return trace.default_process_primitive(outside_call_p, args, params)
-  # Split into two taps, one for the knowns and one for the unknowns
-  # We implement here only the case when primals are known, and we make a tap
-  # with just the primals.
   primals, tangents = util.split_list(args, [nr_primals])
-  c_primals_tapped, _ = util.split_list(consts, [nr_primals])
-  assert all([c is not None for c in c_primals_tapped])
+  all_primals_known = all(p.is_known() for p in primals)
+  some_tangents_unknown = any(not t.is_known() for t in tangents)
+
+  if not (all_primals_known and some_tangents_unknown):
+    return trace.default_process_primitive(outside_call_p, args, params)
 
   prims, _ = params["arg_treedef"].unflatten(args)
   _, primals_treedef = api.tree_flatten(prims)

--- a/jax/interpreters/partial_eval.py
+++ b/jax/interpreters/partial_eval.py
@@ -147,7 +147,7 @@ class JaxprTrace(Trace):
 
   def default_process_primitive(self, primitive, tracers, params):
     """By default, if all the input tracers are known, then execute the primitive
-    and all the ouputs are known. Otherwise, all the outputs are unknown."""
+    and all the outputs are known. Otherwise, all the outputs are unknown."""
     consts = [t.pval.get_known() for t in tracers]
     if all(c is not None for c in consts):
       return primitive.bind(*consts, **params)

--- a/tests/host_callback_test.py
+++ b/tests/host_callback_test.py
@@ -1841,7 +1841,7 @@ class HostCallbackTapTest(jtu.JaxTestCase):
 
     jax.grad(loss)(1.0)  # should not fail
 
-  def test_tap_remat(self):
+  def test_tap_remat_0(self):
     def f(i, k):
       x = hcb.id_print(k + i, output_stream=testing_stream)
       return k * x
@@ -1855,6 +1855,47 @@ class HostCallbackTapTest(jtu.JaxTestCase):
       3
       10"""
     self.assertMultiLineStrippedEqual(expected, testing_stream.output)
+
+  @parameterized.named_parameters(
+      jtu.cases_from_list(
+          dict(testcase_name=f"_use_remat={use_remat}_{grad_func}_use_result={use_result}",
+               use_result=use_result, use_remat=use_remat, grad_func=grad_func)
+          for use_result in [True, False]
+          for grad_func in ["grad", "value_and_grad"]
+          for use_remat in [True, False]))
+  def test_tap_remat(self, use_result=False, grad_func="grad", use_remat=False):
+    def f(x):
+      id_print_result = hcb.id_print(x, output_stream=testing_stream)
+      if use_result:
+        x = id_print_result
+      return 3. * x
+    grad_f = jax.grad if grad_func == "grad" else jax.value_and_grad
+    trans_f = jax.remat(f) if use_remat else f
+    print(jax.make_jaxpr(grad_f(trans_f))(2.))
+    grad_f(trans_f)(2.)
+
+    hcb.barrier_wait()
+
+    if not use_result:
+      if use_remat:
+        expected = ""
+      else:
+        # TODO: if not use_result then we should only see the primal when
+        # computing value_and_grad.
+        expected = "2."
+    elif use_remat:
+      expected = """
+        2.
+        2.
+        transforms: ['jvp', 'transpose']
+        3."""
+    else:
+      expected = """
+        2.
+        transforms: ['jvp', 'transpose']
+        3."""
+    self.assertMultiLineStrippedEqual(expected, testing_stream.output)
+    testing_stream.reset()
 
   def test_tap_named_call(self):
     def tap_scalar(init, do_print=False):

